stan/math/rev/mat/functor/adj_jac_apply.hpp

```diff
@@ -48,6 +48,106 @@ template <class F, class Tuple>
 constexpr auto apply(const F& f, const Tuple& t) {
   return apply_impl(f, t, std::make_index_sequence<std::tuple_size<Tuple>{}>{});
 }
+
+/**
+ * build_y_adj takes the adjoint from the vari pointed to
+ * by y_vi_[0] and stores it in y_adj
+ *
+ * @tparam size dimensionality of M_
+ * @param y_vi_ pointer to pointer to vari
+ * @param M_ ignored in this specialization
+ * @param[out] y_adj reference to variable where adjoint is to be stored
+ */
+template <size_t size>
+void build_y_adj(vari** y_vi_, const std::array<int, size>& M_, double& y_adj) {
+  y_adj = y_vi_[0]->adj_;
+}
+
+/**
+ * build_y_adj takes the adjoints from the varis pointed to
+ * by y_vi_ and stores them in y_adj
+ *
+ * @tparam size dimensionality of M_
+ * @param y_vi_ pointer to pointers to varis
+ * @param M_ shape of y_adj
+ * @param[out] y_adj reference to std::vector where adjoints are to be stored
+ */
+template <size_t size>
+void build_y_adj(vari** y_vi_, const std::array<int, size>& M_,
+                 std::vector<double>& y_adj) {
+  y_adj.resize(M_[0]);
+  for (int m = 0; m < M_[0]; ++m)
```

Comment:
[for the future]
We should really have a routine that copies adjoints of one iterator into another to remove all this redundant code.  We already have `value_of`, but we want `adjoints_of` and her we want to copy into an existing collection.